services/beaconchain/src/main/java/tech/pegasys/artemis/services/beaconchain/BeaconChainController.java

```diff
@@ -271,48 +269,49 @@ private void onTick(Date date) {
     if (!testMode && !stateProcessor.isGenesisReady()) {
       return;
     }
+    final UnsignedLong currentTime = UnsignedLong.valueOf(date.getTime() / 1000);
+    if (chainStorageClient.getStore() != null) {
+      final Store.Transaction transaction = chainStorageClient.getStore().startTransaction();
+      on_tick(transaction, currentTime);
+      chainStorageClient.commit(transaction).join();
+      final UnsignedLong nextSlotStartTime =
+          chainStorageClient
+              .getGenesisTime()
+              .plus(nodeSlot.times(UnsignedLong.valueOf(SECONDS_PER_SLOT)));
+      if (chainStorageClient.getStore().getTime().compareTo(nextSlotStartTime) >= 0) {
+        processSlot();
+      }
+    }
+  }
+
+  private void processSlot() {
```

Comment:
I think it would make sense here to make onTick and processSlot synchronized. I can imagine a scenario in which the timer that onTick listens to sends two Date objects one soon after another and before the initial thread gets to the end of processSlot and update the nodeSlot variable, the second thread gets in and thinks we're already at the previous slot. In that scenario our nodeSlot would be increased by both threads, where we'd only want one of them to be successful increasing the node slot.

Reply:
`processSlot` is only ever called from `onTick` so only one would need to be synchronised.  because of the way the `AsyncEventBus` works, because `onTick` doesn't explicitly have `@AllowConcurrentEvents` annotation, it will only ever be delivered one event at a time (to this method) so it is effectively synchronised anyway.  We wind up depending on that behaviour of `EventBus` in a few places already. It makes me somewhat uncomfortable but if we're going to change I'd be inclined to really think through threading rather than just address it piecemeal.